### DISCUSSION TITLE
ParamSpec Question

### DISCUSSION THREAD

**Discussion by max-muoto** [microsoft/pyright#6816] [General]
Was curious if anyone had ideas on this.

```python
from typing import Callable, ParamSpec, TypeVar

P = ParamSpec("P")
R = TypeVar("R")


def log_to_database() -> None:
    ...


def add_logging(f: Callable[P, R]) -> Callable[P, R]:
    def inner(*args: P.args, **kwargs: P.kwargs) -> R:
        log_to_database()
        ex = kwargs["test"] 
        print(typing.reveal_type(ex))  # Typed as `builtins .object`
        return f(*args, **kwargs)

    return inner
```

When using `ParamSpec`, it seems that values pulled from `kwargs` are always as typed as `object` by Pyright and MyPy. Is there any way to provide additional typing information to avoid this? Or is there any reason these aren't just typed as `Any`?

**Comment by erictraut** [comment]
The most that a type checker can say about `kwargs` is that it's a `Mapping[str, object]`. This is slightly better than `Mapping[str, Unknown]`.

If you're using a `ParamSpec`, it's dangerous to assume anything about the captured arguments. The arguments are meant to be passed along to a function that is known to accept that signature, so you should generally treat `P.args` and `P.kwargs` as a black box.

**Reply by max-muoto** [reply]
Thanks, makes sense!
